conductor_api/src/interface.rs

```diff
@@ -740,12 +741,30 @@ impl ConductorApiBuilder {
         self
     }
 
-    pub fn with_mock_signing_callback(&mut self) {
+    pub fn with_agent_signature_callback(mut self, keypair: Arc<Mutex<Keypair>>) -> Self {
         self.io.add_method("agent/sign", move |params| {
             let params_map = Self::unwrap_params_map(params)?;
             let payload = Self::get_as_string("payload", &params_map)?;
-            Ok(json!({"payload": payload, "signature": mock_signer(payload)}))
+            // Convert payload string into a SecBuf
+            let mut message = SecBuf::with_insecure_from_string(payload.clone());
+            // Create signature
+            let mut message_signed = SecBuf::with_insecure(SIGNATURESIZE);
```

Comment:
I just got to look at this stuff for the first time. Are you concerned that mutation is creeping into higher layers of abstraction? 

Is there any reason why we need an `Arc<Mutex<_>>` for any reason other than that keypair.sign requires does in-place mutation of its arguments? Is there any reason why we can't make keypair::sign create its own empty buffer and return that instead of passing one in from outside? Wouldn't that remove the need for mutable references and maybe this mutex?

Reply:
I looked at it more and it does look like there's no way around this, based on how holochain-sodium is implemented. I know this is a little out of scope for this PR but I'm wondering why we need all this shared memory and locking stuff for signing, and if there's any way to keep it from infecting the higher abstractions?

Reply:
We need that because in order to sign we need the agent's private key. `Keypair` is wrapping the key, stored in secure memory, so that we can use keys without having to deal with memory locking etc. and still have a secure way to handle keys. The only drawback is this: we need have a mutable reference to it in order to use the key, because we need to make the memory readable for us. And yes, that is why I switched to `Arc<Mutex<Keypair>>` here..

Reply:
I see, it's because of secure memory. If this gets out of hand though, we might want to find a way to contain this shared mutation in an abstraction so it doesn't run amok and infect everything.

Reply:
Indeed there's no reason we can't implement internal mutability to these sodium functions to achieve signing and other crypto on with read-only keys (although we'd probably want to use an unsafe variant so we don't get runtime rust panics) . Just need to have a chance to catch our breaths and prioritize stuff like that...